### PR TITLE
Small Chosen Tweaks

### PR DESCRIPTION
**Pull request by Grobobobo**

### CODE REVIEW

**Comment by pledbrook** [comment]
I suggest looking at RPGO's `X2Effect_CapStat` if you really want Unstoppable in the game. I'm not sure it's devoid of its own problems, but I can't see any obvious issues.

**Comment by Grobobobo** [comment]
Well, there aren't any percentage mobility reductions the player can use against ass, so the effect works in it's very narrow intended use. But yeah, if you want a more general purpose solution then I'll switch to the RPGO one.

File: LongWarOfTheChosen/Src/LW_Overhaul/Classes/X2Ability_LW_ChosenAbilities.uc [hunk 1]
```diff
@@ -1823,6 +1830,56 @@ static function X2AbilityTemplate AssassinBladestormAttack()
...
+	local X2AbilityTemplate						Template;	
+	local X2Effect_CapStat 					UnstoppableEffect;
+
+	`CREATE_X2ABILITY_TEMPLATE(Template, 'Unstoppable_LW');
+	Template.IconImage = "img:///UILibrary_XPerkIconPack.UIPerk_move_blaze";
+	Template.AbilitySourceName = 'eAbilitySource_Perk';
+	Template.Hostility = eHostility_Neutral;
+	Template.eAbilityIconBehaviorHUD = eAbilityIconBehavior_NeverShow;
+	Template.AbilityToHitCalc = default.DeadEye;
+    Template.AbilityTargetStyle = default.SelfTarget;
```
Comment (pledbrook) [inline LongWarOfTheChosen/Src/LW_Overhaul/Classes/X2Ability_LW_ChosenAbilities.uc:1844]:
```suggestion
	Template.AbilityTargetStyle = default.SelfTarget;
```

File: LongWarOfTheChosen/Config/XComLW_SoldierSkills.ini [hunk 2]
```diff
@@ -878,6 +878,7 @@ IMPACT_COMPENSATION_MAX_STACKS=3
 
 SHIELD_ALLY_PCT_DR=0.67
 
+UNSTOPPABLE_MIN_MOB = 7
```
Comment (pledbrook) [inline LongWarOfTheChosen/Config/XComLW_SoldierSkills.ini:881]:
```suggestion
UNSTOPPABLE_MIN_MOB = 7

```

File: LongWarOfTheChosen/Src/LW_Overhaul/Classes/X2Ability_LW_ChosenAbilities.uc [hunk 3]
```diff
@@ -1823,6 +1830,56 @@ static function X2AbilityTemplate AssassinBladestormAttack()
...
+	Template.bDisplayInUITooltip = true;
+	Template.bDisplayInUITacticalText = true;
+	Template.AbilityShooterConditions.AddItem(default.LivingShooterProperty);
+	Template.AbilityTriggers.AddItem(default.UnitPostBeginPlayTrigger);
+
+
+	UnstoppableEffect = new class'X2Effect_CapStat';
+	UnstoppableEffect.BuildPersistentEffect(1,false,true,,eGameRule_PlayerTurnBegin);
+	UnstoppableEffect.SetDisplayInfo(ePerkBuff_Bonus, Template.LocFriendlyName, Template.GetMyLongDescription(), Template.IconImage, false,,Template.AbilitySourceName);
+	UnstoppableEffect.AddStatCap(eStat_Mobility,default.UNSTOPPABLE_MIN_MOB,true);
```
Comment (pledbrook) [inline LongWarOfTheChosen/Src/LW_Overhaul/Classes/X2Ability_LW_ChosenAbilities.uc:1857]:
```suggestion
	UnstoppableEffect.BuildPersistentEffect(1, false, true,, eGameRule_PlayerTurnBegin);
	UnstoppableEffect.SetDisplayInfo(ePerkBuff_Bonus, Template.LocFriendlyName, Template.GetMyLongDescription(), Template.IconImage, false,, Template.AbilitySourceName);
	UnstoppableEffect.AddStatCap(eStat_Mobility,default.UNSTOPPABLE_MIN_MOB,true);
```

File: LongWarOfTheChosen/Src/LW_Overhaul/Classes/X2Effect_CapStat.uc [hunk 4]
```diff
@@ -0,0 +1,163 @@
...
+
+class X2Effect_CapStat extends X2Effect_PersistentStatChange;
+
+var array<X2Condition> Conditions;
+
+struct StatCap
+{
+	var ECharStatType StatType;
+	var float StatCapValue;
+	var bool Minimum; // If true, the cap means that the stat cannot go lower than the amount, if false, the higher
```
Comment (pledbrook) [inline LongWarOfTheChosen/Src/LW_Overhaul/Classes/X2Effect_CapStat.uc:15]:
```suggestion
	var bool IsMinimum; // If true, the cap means that the stat cannot go lower than the amount, if false, the higher
```

File: LongWarOfTheChosen/Src/LW_Overhaul/Classes/X2Effect_CapStat.uc [hunk 5]
```diff
@@ -0,0 +1,163 @@
...
+};
+
+var array<StatCap> StatCaps; 
+
+function AddStatCap(ECharStatType StatType, float StatCapValue, bool Minimum)
```
Comment (pledbrook) [inline LongWarOfTheChosen/Src/LW_Overhaul/Classes/X2Effect_CapStat.uc:20]:
```suggestion
function AddStatCap(ECharStatType StatType, float StatCapValue, bool IsMinimum)
```

File: LongWarOfTheChosen/Src/LW_Overhaul/Classes/X2Effect_CapStat.uc [hunk 6]
```diff
@@ -0,0 +1,163 @@
...
+{
+	local StatCap Cap;
+	
+	Cap.StatType = StatType;
+	Cap.StatCapValue = StatCapValue;
+	Cap.Minimum = Minimum;
```
Comment (pledbrook) [inline LongWarOfTheChosen/Src/LW_Overhaul/Classes/X2Effect_CapStat.uc:26]:
```suggestion
	Cap.IsMinimum = IsMinimum;
```

File: LongWarOfTheChosen/Src/LW_Overhaul/Classes/X2Effect_CapStat.uc [hunk 7]
```diff
@@ -0,0 +1,163 @@
...
+	local XComGameState NewGameState;
+	local X2Effect_CapStat EffectTemplate;
+	local XComGameState_Effect_CapStats EffectState;
+	local bool bOldApplicable, bNewApplicable;
+	local array<StatChange> LocalStatChanges;
+	local StatChange LocalStatChange;
+	local StatCap LocalStatCap;
+	local int AppliedStatChangeIndex, StatChangeOther;
+	local float AppliedStatChange, CappedStat, NewStatAmount;
+	local UnitValue ImmobilizeValue;
```
Comment (pledbrook) [inline LongWarOfTheChosen/Src/LW_Overhaul/Classes/X2Effect_CapStat.uc:63]:
```suggestion
	local UnitValue ImmobilizeValue;

```

File: LongWarOfTheChosen/Src/LW_Overhaul/Classes/X2Effect_CapStat.uc [hunk 8]
```diff
@@ -0,0 +1,163 @@
...
+	EventMgr = `XEVENTMGR;
+
+	UnitState = XComGameState_Unit(`XCOMHISTORY.GetGameStateForObjectID(EffectGameState.ApplyEffectParameters.TargetStateObjectRef.ObjectID));
+
+	ListenerObj = EffectGameState;
+
+	// Register to tick after EVERY action.
+	EventMgr.RegisterForEvent(ListenerObj, 'OnUnitBeginPlay', EventHandler, ELD_OnStateSubmitted, 25, UnitState,, EffectGameState);	
+	//EventMgr.RegisterForEvent(ListenerObj, 'AbilityActivated', EventHandler, ELD_OnStateSubmitted, 25,,, EffectGameState);
+	EventMgr.RegisterForEvent(ListenerObj, 'UnitTakeEffectDamage', EventHandler, ELD_OnStateSubmitted, 25,,, EffectGameState);
```
Comment (pledbrook) [inline LongWarOfTheChosen/Src/LW_Overhaul/Classes/X2Effect_CapStat.uc:45]: Why did you change this? `UnitTakeEffectDamage` makes this approach even less reliable, since it will never proc for abilities that don't damage but do apply stat buffs/debuffs. This was the whole reason for me suggesting Musashi's implementation.

**Review by pledbrook — CHANGES_REQUESTED** [review]
You haven't addressed my previous concern about Unstoppable. The main reason for suggesting Musashi's implementation, even though I think that approach is still problematic, is because it's based around `AbilityActivated`.
